indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java

```diff
@@ -172,4 +197,232 @@ public static void setInputFormat(Job job, HadoopDruidIndexerConfig indexerConfi
+                  descriptorPath,
+                  EnumSet.of(CreateFlag.OVERWRITE, CreateFlag.CREATE),
+                  Options.CreateOpts.bufferSize(256 * 1024),
+                  Options.CreateOpts.createParent()
+              )) {
+                HadoopDruidIndexerConfig.jsonMapper.writeValue(descriptorOut, segment);
+                descriptorOut.flush();
+              }
+            }
+            catch (RuntimeException | IOException ex) {
+              log.info(ex, "Error in retry loop");
+              throw ex;
+            }
+          }
+        },
+        RetryPolicies.retryUpToMaximumCountWithFixedSleep(NUM_RETRIES, SECONDS_BETWEEN_RETRIES, TimeUnit.SECONDS)
+    );
+    descriptorPusher.push();
+  }
+
+  /**
+   * Simple interface for retry operations
+   */
+  public interface DataPusher
+  {
+    void push() throws IOException;
+  }
+
+  public static long zipAndCopyDir(
+      File baseDir,
+      OutputStream baseOutputStream,
+      Progressable progressable
+  ) throws IOException
+  {
+    long size = 0L;
+    try (ZipOutputStream outputStream = new ZipOutputStream(baseOutputStream)) {
+      List<String> filesToCopy = Arrays.asList(baseDir.list());
+      for (String fileName : filesToCopy) {
+        final File fileToCopy = new File(baseDir, fileName);
+        if (java.nio.file.Files.isRegularFile(fileToCopy.toPath())) {
+          size += copyFileToZipStream(fileToCopy, outputStream, progressable);
+        } else {
+          log.warn("File at [%s] is not a regular file! skipping as part of zip", fileToCopy.getPath());
+        }
+      }
+      outputStream.flush();
+    }
+    return size;
+  }
+
+  public static long copyFileToZipStream(
+      File file,
+      ZipOutputStream zipOutputStream,
+      Progressable progressable
+  ) throws IOException
+  {
+    createNewZipEntry(zipOutputStream, file);
+    long numRead = 0;
+    try (FileInputStream inputStream = new FileInputStream(file)) {
+      byte[] buf = new byte[0x10000];
+      for (int bytesRead = inputStream.read(buf); bytesRead >= 0; bytesRead = inputStream.read(buf)) {
+        progressable.progress();
+        if (bytesRead == 0) {
+          continue;
+        }
+        zipOutputStream.write(buf, 0, bytesRead);
+        progressable.progress();
+        numRead += bytesRead;
+      }
+    }
+    zipOutputStream.closeEntry();
+    progressable.progress();
+    return numRead;
+  }
+
+  private static void createNewZipEntry(ZipOutputStream out, File file) throws IOException
+  {
+    log.info("Creating new ZipEntry[%s]", file.getName());
+    out.putNextEntry(new ZipEntry(file.getName()));
+  }
+
+  public static Path makeSegmentOutputPath(
+      Path basePath,
+      FileSystem fileSystem,
+      String dataSource,
+      String version,
+      Interval interval,
+      int partitionNum
+  )
+  {
+    Path outputPath = new Path(prependFSIfNullScheme(fileSystem, basePath), "./" + dataSource);
```

Comment:
They are required due to Path's interpretation of "something:otherthing" as an absolute instead of relative string, so Path(somePath, "something:otherthing") will cause problems due to the later argument being an 'absolute' path instead of a relative one. But Path(somePath, "./something:otherthing") is fine